In [15]:
# packages

import math
import pandas as pd
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
#nltk.download('punkt')
#nltk.download('stopwords')

In [17]:
# creating the dataframe with original skincare products dataset

products = pd.read_csv('data/skincare_products.csv')

In [18]:
def cosine_similarity(X, Y):

  # establish stop words
  X_list = word_tokenize(X)
  Y_list = word_tokenize(Y)
  sw = stopwords.words('english')
  l1 =[];l2 =[]

  # remove stop words from strings
  X_set = {w for w in X_list if not w in sw}
  Y_set = {w for w in Y_list if not w in sw}

  # form set with words from both strings
  rvector = X_set.union(Y_set)
  for w in rvector:
    if w in X_set: l1.append(1) # create a vector
    else: l1.append(0)
    if w in Y_set: l2.append(1)
    else: l2.append(0)
  c = 0

  # cosine similarity formula
  for i in range(len(rvector)):
    c+= l1[i]*l2[i]
    cosine = c / float((sum(l1)*sum(l2))**0.5)

  return cosine

In [20]:
products

Label                          Brand  \
0     Moisturizer                         LA MER   
1     Moisturizer                          SK-II   
2     Moisturizer                 DRUNK ELEPHANT   
3     Moisturizer                         LA MER   
4     Moisturizer                   IT COSMETICS   
...           ...                            ...   
1467  Sun protect                         KORRES   
1468  Sun protect                KATE SOMERVILLE   
1469  Sun protect                  VITA LIBERATA   
1470  Sun protect  ST. TROPEZ TANNING ESSENTIALS   
1471  Sun protect                     DERMAFLASH   

                                                   Name  Price  Rank  \
0                                       Crème de la Mer    175   4.1   
1                              Facial Treatment Essence    179   4.1   
2                            Protini™ Polypeptide Cream     68   4.4   
3                           The Moisturizing Soft Cream    175   3.8   
4         Your Skin But Better™ CC+™ Cream with SPF 50+     38   4.1   
...                                                 ...    ...   ...   
1467  Yoghurt Nourishing Fluid Veil Face Sunscreen B...     35   3.9   
1468  Daily Deflector™ Waterlight Broad Spectrum SPF...     48   3.6   
1469                            Self Tan Dry Oil SPF 50     54   3.5   
1470                   Pro Light Self Tan Bronzing Mist     20   1.0   
1471  DERMAPROTECT Daily Defense Broad Spectrum SPF 50+     45   0.0   

                                            Ingredients  Combination  Dry  \
0     Algae (Seaweed) Extract, Mineral Oil, Petrolat...            1    1   
1     Galactomyces Ferment Filtrate (Pitera), Butyle...            1    1   
2     Water, Dicaprylyl Carbonate, Glycerin, Ceteary...            1    1   
3     Algae (Seaweed) Extract, Cyclopentasiloxane, P...            1    1   
4     Water, Snail Secretion Filtrate, Phenyl Trimet...            1    1   
...                                                 ...          ...  ...   
1467  Water, Alcohol Denat., Potassium Cetyl Phospha...            1    1   
1468  Water, Isododecane, Dimethicone, Butyloctyl Sa...            0    0   
1469  Water, Dihydroxyacetone, Glycerin, Sclerocarya...            0    0   
1470  Water, Dihydroxyacetone, Propylene Glycol, PPG...            0    0   
1471                      Visit the DERMAFLASH boutique            1    1   

      Normal  Oily  Sensitive  
0          1     1          1  
1          1     1          1  
2          1     1          0  
3          1     1          1  
4          1     1          1  
...      ...   ...        ...  
1467       1     1          1  
1468       0     0          0  
1469       0     0          0  
1470       0     0          0  
1471       1     1          1  

[1472 rows x 11 columns]

In [19]:
# testing cosine_similarity function on products in database

# comparing first and second products in the skincare products database
test = cosine_similarity(products.iloc[0, 5], products.iloc[1, 5])
print(test)

# comparing first and third products in the skincare products database
test2 = cosine_similarity(products.iloc[0, 5], products.iloc[2, 5])
print(test2)

# comparing first and fourth products in the skincare products database
test2 = cosine_similarity(products.iloc[0, 5], products.iloc[3, 5])
print(test2)

print('Out of these comparisons, products 1 and 4 are most similar with this method')

0.22555354977384037
0.1487044791289829
0.5283120976241462
Out of these comparisons, products 1 and 4 are most similar with this method


* Create a 1472*1472 matrix of all of the similarity scores, i.e., the entry in row 1 column 2 should contain the cosine similarity between product 1 and product 2 